### DISCUSSION TITLE
The naming convention behind _def and _ctx

### DISCUSSION THREAD

**Discussion by lukaszmakuch** [lukemorales/query-key-factory#64] [Q&A] [answered]
Hi! 👋

I have a question about the naming convention behind fields such as `_def` and `_ctx`.
When I first saw them, I thought they were meant to be private.
But I concluded that they are actually public after reading the documentation and seeing them being used in application code, such as here:
```js
      queryClient.invalidateQueries({
        queryKey: queries.todos.list._def,
        refetchActive: false,
      });
```
and here
```
export function useUserLikes(userId: string) {
  return useQuery(users.detail(userId)._ctx.likes);
};
```

**My question is: are `_def` and `_ctx` public fields that I can safely use in my code, or are they private fields I shouldn't touch?**

Cheers!

**Comment by lukemorales** [comment, marked as answer]
Hi @lukaszmakuch everything exposed for you through the typesystem is supposed to be used. The only reason they start with `_` is to differentiate that these are "internal" properties auto generated by the lib, and not something yourself has declared as an entry key (if you try to create a key starting with `_` you will have a runtime error), and also to not create a deny list of common property names you can't use.
These properties serve to give you access to some features like the query definition `_def` e.g., the fundamental query key to a composable query key (the `_def` of `['users', { userId: 1 }]` is `['users']`), or to give you access to contextual queries like shown in the docs.
